RFC-0007-remote-functions.md

```diff
@@ -0,0 +1,219 @@
+Presto C++ deployments, and will unify function server implementations under one API.
+
+In the future, a plugin concept for remote functions should be added to Presto C++ to allow for custom function server implementations,
+however this is out of scope for this RFC.
+
+![REST API for remote functions](RFC-0007/Diagram.png)
+
+### Functions API
+
+An OpenAPI specification will be created which defines a REST API for remote function servers.  This API will feature the following
+endpoints:
+
+#### Functions version
+
+> Endpoint: /v1/functions
+>
+> HTTP verb: HEAD
+>
+> Request body: empty
+>
+> Response body: empty
+
+Returns the headers of the GET response.  This is useful for checking the version of the API, which will be returned as a header.
+
+#### List all functions
+
+> Endpoint: /v1/functions
+>
+> HTTP verb: GET
+>
+> Request body: empty
+>
+> Response body: JSON array of function metadata objects
+
+Returns the complete listing of functions across all schemas.
+
+#### List functions at schema
+
+> Endpoint: /v1/functions/{schema}
+>
+> HTTP verb: GET
+>
+> Request body: empty
+>
+> Response body: JSON array of function metadata objects
+
+Returns the complete listing of all functions in the specified schema.
+
+#### List functions with at schema with name
+
+> Endpoint: /v1/functions/{schema}/{functionName}
+>
+> HTTP verb: GET
+>
+> Request body: empty
+>
+> Response body: JSON array of function metadata objects
+
+Returns the complete listing of functions in the specified schema with the specified function name.
+
+#### Add a function
+
+> Endpoint: /v1/functions/{schema}/{functionName}
+>
+> HTTP verb: POST
+>
+> Request body: JSON object representing the function to be added
+>
+> Response body: the function ID of the newly created function
+
+Creates a new function in the specified schema with the specified name.  The function object will contain the metadata of the
+function, including its arguments, return type, and other metadata.  It will return an identifier representing this
+specific function, which is useful to differentiate multiple functions which share the same name but have different
+arguments.
+
+#### Update a function
+
+> Endpoint: /v1/functions/{schema}/{functionName}/{functionId}
+>
+> HTTP verb: PUT
+>
+> Request body: JSON object representing the function to be updated
+>
+> Response body: the function ID of the newly created function
+
+Updates the function in the specified schema with the specified name and function ID.
+
+#### Delete a function
+
+> Endpoint: /v1/functions/{schema}/{functionName}/{functionId}
+>
+> HTTP verb: DELETE
+>
+> Request body: empty
+>
+> Response body: empty
+
+Deletes the function in the specified schema with the specified name and function ID.
+
+#### Execute a function
+
+> Endpoint: /v1/functions/{schema}/{functionName}/{functionId}/{version}
+>
+> HTTP verb: POST
+>
+> Request body: Presto Page of input data
```

Comment:
Can we also support arrow here?

Reply:
Nvm its just bytes

Reply:
I will update the spec to show that Arrow is an acceptable option.

Reply:
Added a sentence describing initial support for Presto Page but later support for Arrow.